### PR TITLE
mobile: add Swift C++ interop implementation

### PR DESCRIPTION
**Pull request by jpsim**
This acts as a drop-in replacement for the existing Objective-C implementation that we've always used for the Swift APIs.

It leverages [Swift's experimental C++ interoperability support](https://github.com/apple/swift/blob/main/docs/CppInteroperability/CppInteroperabilityStatus.md) to allow us to write Swift code that interacts directly with Envoy's C++ APIs.

Also add a new Swift C++ interop test suite.

This is still unreachable from the main library's API, so it's very low risk as-is, but that will be hooked up in an upcoming change.

~Depends on https://github.com/envoyproxy/envoy/pull/25971~

Commit Message:
Additional Description:
Risk Level: Low, unreachable code
Testing: Added new test suite with tests for much of the new code.
Docs Changes: None, unreachable code
Release Notes: None, unreachable code
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: mobile/library/swift/cxx/EnvoyHTTPFilterFactory+Cxx.swift [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+      let responseFilterCallbacks = EnvoyHTTPFilterCallbacksImpl(callbacks: callbacks)
+      setResponseFilterCallbacks(responseFilterCallbacks)
+    }
+
+    filter.pointee.on_resume_request =
+    { headers, data, trailers, end_stream, stream_intel, context in
+      let filter = Unmanaged<EnvoyHTTPFilter>.fromOpaque(context!).takeUnretainedValue()
+      guard let onResumeRequest = filter.onResumeRequest else {
+        return envoy_filter_resume_status(
+          status: kEnvoyFilterHeadersStatusContinue,
```
Comment (Augustyniak) [inline mobile/library/swift/cxx/EnvoyHTTPFilterFactory+Cxx.swift:191]: This should be kEnvoyFilterResumeStatusResumeIteration

File: mobile/library/swift/cxx/EnvoyHTTPFilterFactory+Cxx.swift [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        pending_trailers: toEnvoyHeadersPtr(result[3] as? [String: [String]])
+      )
+    }
+
+    filter.pointee.on_resume_response =
+    { headers, data, trailers, end_stream, stream_intel, context in
+      let filter = Unmanaged<EnvoyHTTPFilter>.fromOpaque(context!).takeUnretainedValue()
+      guard let onResumeResponse = filter.onResumeResponse else {
+        return envoy_filter_resume_status(
+          status: kEnvoyFilterHeadersStatusContinue,
```
Comment (Augustyniak) [inline mobile/library/swift/cxx/EnvoyHTTPFilterFactory+Cxx.swift:215]: shouldn't it be kEnvoyFilterResumeStatusResumeIteration?

File: mobile/library/swift/cxx/EnvoyHTTPFilterFactory+Cxx.swift [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+          pending_headers: headers,
+          pending_data: data,
+          pending_trailers: trailers
+        )
+      }
+      let pendingHeaders = (headers?.pointee).flatMap(toSwiftHeaders)
+      let pendingData = (data?.pointee).flatMap(Data.init)
```
Comment (Augustyniak) [inline mobile/library/swift/cxx/EnvoyHTTPFilterFactory+Cxx.swift:222]: obj-c implementation uses `release_envoy_data` which frees memory allocated by data. We are leaking this memory in the new implementation I think.

File: mobile/library/swift/cxx/EnvoyHeaders.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    if
+      let headerKey = String.fromEnvoyData(header.key),
+      let headerValue = String.fromEnvoyData(header.value)
+    {
+      result[headerKey, default: []].append(headerValue)
+    }
+  }
+}
+
+func toEnvoyHeaders(_ headers: EnvoyHeaders?) -> envoy_headers {
```
Comment (Augustyniak) [inline mobile/library/swift/cxx/EnvoyHeaders.swift:21]: I think that our current naming scheme is to call these to toNativeHeaders (or swift_toNativeHeaders if there is name collision). What do you think?

File: mobile/library/swift/cxx/EnvoyLoggerFactory.swift [hunk 5]
```diff
@@ -0,0 +1,29 @@
+@_implementationOnly import EnvoyCxxSwiftInterop
+
+// swiftlint:disable force_unwrapping
+
+private typealias Logger = ((String) -> Void)?
+private typealias BoxType = PointerBox<Logger>
+
+enum EnvoyLoggerFactory {
```
Comment (Augustyniak) [inline mobile/library/swift/cxx/EnvoyLoggerFactory.swift:8]: doc string, add a note about what happens when passed log is nil

File: mobile/library/swift/cxx/EnvoyLoggerFactory.swift [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  static func create(log: ((String) -> Void)?) -> envoy_logger {
+    guard let log else {
+      return envoy_logger(log: nil, release: nil, context: nil)
+    }
+
+    return envoy_logger(
+      log: { stringData, context in
+        guard let string = String.fromEnvoyData(stringData) else {
```
Comment (Augustyniak) [inline mobile/library/swift/cxx/EnvoyLoggerFactory.swift:16]: I think that obj-c implementation has autorelease pool in here

File: mobile/library/swift/cxx/EnvoyEventTrackerFactory.swift [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+enum EnvoyEventTrackerFactory {
+  static func create(track: (([String: String]) -> Void)?) -> envoy_event_tracker {
+    guard let track else {
+      return envoy_event_tracker(track: nil, context: nil)
+    }
+
+    // TODO(jpsim): Add `release` field
+    return envoy_event_tracker(
+      track: { map, context in
+        let mutablePointer = UnsafeMutableRawPointer(mutating: context!)
```
Comment (Augustyniak) [inline mobile/library/swift/cxx/EnvoyEventTrackerFactory.swift:17]: I think that obj-c implementation has autorelease pool in here. I think that we need to add it here as well

File: mobile/library/swift/cxx/EnvoyStringAccessor+Cxx.swift [hunk 8]
```diff
@@ -0,0 +1,18 @@
+@_implementationOnly import EnvoyCxxSwiftInterop
+@_implementationOnly import EnvoyEngine
+
+// swiftlint:disable force_cast force_unwrapping
+
+extension EnvoyStringAccessor {
```
Comment (Augustyniak) [inline mobile/library/swift/cxx/EnvoyStringAccessor+Cxx.swift:6]: let's add some basic doc string

**Review by Augustyniak — COMMENTED** [review]
I reviewed a portion (maybe ~50%) of files in this PR. Submitting my partial PR since it's better than nothing and I need to move to other efforts for the next few hours. Curious of your comments to some of my memory (autorelease pool)/flow (calling callback with continue as opposed to resume) related comments. 

For the record, I do think that this is a high-risk change.

**Comment by jpsim** [comment]
> I reviewed a portion (maybe ~50%) of files in this PR. Submitting my partial PR since it's better than nothing and I need to move to other efforts for the next few hours.

I appreciate you taking a look!

> Curious of your comments to some of my memory (autorelease pool)/flow (calling callback with continue as opposed to resume) related comments.

I'll be working through your comments with a focus on the memory management.

At a high-level I will say that I measured memory usage in all our example apps using Instruments & Xcode's memory graph debugger comparing both the objc (`useSwiftCxxInterop(false)`) and the swift (`useSwiftCxxInterop(true)`) implementations and they were in the same ballpark, and when terminating & deallocating the engine, nothing relevant leaked in either implementation.

> For the record, I do think that this is a high-risk change.

Maybe I'm being pedantic, but this change is low risk as-is, since all this is new code that should be 100% unreachable in its current form. That said, the code is also useless in its current form for the exact same reason. 🤷 

I'd characterize https://github.com/envoyproxy/envoy/pull/26011 as medium-high risk since it actually makes this code reachable to the Swift APIs, unless disabled by setting `useSwiftCxxInterop(false)` or building with `--define=envoy_mobile_swift_cxx_interop=disabled`. [This comment](https://github.com/envoyproxy/envoy/pull/26011#issuecomment-1462425786) attempts to characterize the risk factors involved, and I do think that memory management is one of the more likely risks, despite my best attempts to hunt for and catch leaks.

**Comment by Augustyniak** [comment]
>Maybe I'm being pedantic, but this change is low risk as-is, since all this is new code that should be 100% unreachable in its current form. That said, the code is also useless in its current form for the exact same reason. 🤷

Yes, of course 🤓 . My high-risk comment applied for when this feature (Swift implementation) was enabled/active. This is when I think it becomes high-risk. Otherwise it's a low-risk.

File: mobile/library/swift/cxx/Bootstrap.swift [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  init(ptr: Envoy.CxxSwift.BootstrapPtr) {
+    self.ptr = ptr
+  }
+
+  var debugDescription: String {
+    .fromCXX(Envoy.CxxSwift.bootstrapDebugDescription(self.ptr))
+  }
+}
+
+extension Envoy.Platform.EngineBuilder {
```
[the remark this answers is not in the document]
Comment (jpsim) [reply mobile/library/swift/cxx/Bootstrap.swift:15]: I prefer keeping as much of the C++ interop stuff in `library/swift/cxx` so we don't need to guard for `#if canImport(EnvoyCxxSwiftInterop)` and it's easier to remove if we decide not to move forward with it.

**Comment by jpsim** [comment]
Went with https://github.com/envoyproxy/envoy/pull/26111 instead of this as a short-term & low-risk way to dip our toes into Swift / C++ interop waters.

I'll close this PR for now and might revive it after I'm back from my leave in mid-April.

**Pull request closed without merging**